### PR TITLE
Add TransformationTool to transform XML into Text with XSLT file.

### PR DESCRIPTION
**Pull request by manuelsaldivar525**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

(FILL ME IN) This section describes why this PR is here. Usually it would include a reference 
to the tracking task that it is part or all of the solution for.

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

**Comment by mjkkirschner** [comment]
@manuelsaldivar525 please add a description to this PR. It's useful when looking back through the history.

**Comment by manuelsaldivar525** [comment]
Project TransformXMLToJSONUsingXSLT was added under the \Tools, this tool is for get a file in flat text separated by the character ',', this file will obtain the information from the XML file that is the result of process unit tests in the nunit-console in format NUnit2, and the flat text is going to be used for run  parallel tests, you can check the readme file into the project struct to see how to configure this tool.

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+| ------ | ------ | ------ |
+| TransformXMLToText | xslt | File used to transform from XML to Text
+| CustomTextResult | txt | File where results of the convertion will be
+| textFileWithFilters | txt | File where results of the convertion will be but with filters 
+| TestResult-Color_Tests | xml | File´s name can be different because is the file to transform and it is the result of the tests 
+
+Note : The names of the files  can be changed if it is needed.
+
+**Filters**
+This section is the filters that will be applied in case the entry applyFilters has the value 1, in this case the slow tests are the ones that are filtered.
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/README.md]: does this mean these tests will be filtered out or what? What is the point of applying a filter? Please add this to the readme.

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+# Transformation Tool from XML to Text
+
+Transformation Tool is an application to:.
+
+  - Transform files XML to Text
+  - Create text files
+  - Configure the entry files and set the out put files.
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/README.md]: what's an entry file? Whats an output file? in the context of this tool?

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+Where C:\\DSCoreNodesTests.dll is the Dll that contains the set of test, DSCoreNodesTests.ColorTests is the name space of the test and C:\\TestResult-Color_Tests.xml is the path and the name of the file where the result will be.
+
+
+
+### Files needed
+| File | Type | Description
+| ------ | ------ | ------ |
+| TransformXMLToText | xslt | File used to transform from XML to Text
+| CustomTextResult | txt | File where results of the convertion will be
+| textFileWithFilters | txt | File where results of the convertion will be but with filters
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/README.md]: spelling (conversion)

File: src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="JsonXslt, Version=1.0.1.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\JsonXslt.1.0.1\lib\net45\JsonXslt.dll</HintPath>
+    </Reference>
+    <Reference Include="Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj]: can we change this reference to json.net 8.0.3 to align with the rest of dynamo?

File: src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    <OutputPath>bin\Debug\</OutputPath>
+    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
+    <PlatformTarget>AnyCPU</PlatformTarget>
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj]: it looks like this tool builds inside its own folder - which is good! So it should not disturb the dynamo bin folder?

File: src/Tools/TransformXMLToJSONUsingXSLT/XSLT/Transform.xslt [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    </xsl:call-template>
+    
+     <xsl:call-template name="GeneralTemplate">
+      <xsl:with-param  name="prmDiffObject" select="'Class'"/>
+      <xsl:with-param  name="prmLongPath" select="/Assembly/Module/Type/DeclarationDiffs/DiffItem"/>
+    </xsl:call-template>
+    
+    
+
+    <!--<xsl:call-template name="CounterNext">
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/XSLT/Transform.xslt]: whats this commented out stuff?

File: src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj (hunk 4 above)
Comment (mjkkirschner) [reply src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj]: not a big deal if this is coming from the jsonXslt dep and is isolated from Dynamo.

**Comment by mjkkirschner** [comment]
@QilongTang @smangarole @manuelsaldivar525  - a few questions
* is this going to produce any new .dlls in the /bin folder of Dynamo?
* if it does produce new binaries or files in the /bin folder we need to know so we don't effect clients consuming dynamo core runtime nugets which pack the bin folder.
* did this need to be added to the .dynamo .sln?
* should it also be added to the mono.sln?

I'm confused why there are changes to the Dynamo.sln and in addition another .sln being added?

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md (hunk 1 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToJSONUsingXSLT/README.md]: A diagram will be added to explain more about the tool and the context of the same.

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md (hunk 2 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToJSONUsingXSLT/README.md]: A diagram will be added to explain more about the tool and the context of the same.

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md (hunk 3 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToJSONUsingXSLT/README.md]: This mistake will be corrected.

File: src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj (hunk 4 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj]: Yes, this is possible and it is already updated it.

File: src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj (hunk 5 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToJSONUsingXSLT/TransformXMLToJSONUsingXSLT.csproj]: Yes, that´s correct.

File: src/Tools/TransformXMLToJSONUsingXSLT/XSLT/Transform.xslt (hunk 6 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToJSONUsingXSLT/XSLT/Transform.xslt]: This stuff will be removed it.

**Comment by manuelsaldivar525** [comment]
The answers to your questions are:

is this going to produce any new .dlls in the /bin folder of Dynamo?
For the project TransformXMLToJSONUsingXSLT not, this is a .exe.

if it does produce new binaries or files in the /bin folder we need to know so we don't effect clients consuming dynamo core runtime nugets which pack the bin folder.
For the project TransformXMLToJSONUsingXSLT not.

did this need to be added to the .dynamo .sln?
For the project TransformXMLToJSONUsingXSLT has to be added to Dynamo.All sln.

should it also be added to the mono.sln?
For the project TransformXMLToJSONUsingXSLT not.

I hope these answers help.

File: src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs [hunk 7]
```diff
@@ -1,67 +0,0 @@
-﻿using System;
```
Comment (mjkkirschner) [inline src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs]: you do not want to delete this file.... just don't commit any changes to it.

File: src/Dynamo.All.sln [hunk 8]
```diff
@@ -513,6 +513,10 @@ Global
 		{F7FD9395-35D5-474D-8AD1-9904817E70D3}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{F7FD9395-35D5-474D-8AD1-9904817E70D3}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{F7FD9395-35D5-474D-8AD1-9904817E70D3}.Release|Any CPU.Build.0 = Release|Any CPU
+		{676671D5-76B6-4E3B-9244-5323BF6B8EFE}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
Comment (mjkkirschner) [inline src/Dynamo.All.sln]: what are these changes to the dynamo.sln?

Comment (manuelsaldivar525) [reply src/Dynamo.All.sln]: This is because the project TransformXML CSVTool was in the main solution but in the next commit was removed from the main solution and this project TransformXML CSVTool has its own solution on its location.

Comment (mjkkirschner) [reply src/Dynamo.All.sln]: please revert these changes then. Keep the PR as minimal as possible.

File: .gitignore [hunk 9]
```diff
@@ -96,3 +96,4 @@ src/TestResults
 test/core/customast/test.txt
 test/core/files/test.png
 test/core/migration/writetext.txt
+src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs
```
Comment (mjkkirschner) [inline .gitignore]: @manuelsaldivar525 please revert this - this file should not be added to the git ignore - we need to commit it sometimes. Just do not include the auto generated changes in your commit.

File: src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs [hunk 10]
```diff
@@ -1,67 +0,0 @@
...
-using System.Reflection;
```
Comment (mjkkirschner) [inline src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs]: again - do not delete this file - just do not commit the changes.

File: src/Tools/TransformXMLToJSONUsingXSLT/App.config [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    <add key="NodeViewCustomizationTests" value="NodeViewCustomizationTests" />
+    <add key="NodeMigrationTests" value="NodeMigrationTests" />
+    <add key="ComplexTests" value="ComplexTests" />
+    <add key="DSEvaluationModelTest" value="DSEvaluationModelTest" />
+  </Filters>
+  <runtime>
+    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
+      <dependentAssembly>
+        <assemblyIdentity name="Newtonsoft.Json" publicKeyToken="30ad4fe6b2a6aeed" culture="neutral" />
+        <bindingRedirect oldVersion="0.0.0.0-8.0.0.0" newVersion="8.0.0.0" />
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/App.config]: a little odd - I wonder if this occurred when you added the new json 8.0.3 - can you make sure this tool builds and runs fine when you start with a clean bin and packages folder?

File: src/Tools/TransformXMLToJSONUsingXSLT/Properties/AssemblyInfo.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+//
+//      Major Version
+//      Minor Version
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
+[assembly: AssemblyVersion("1.0.0.0")]
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/Properties/AssemblyInfo.cs]: @QilongTang @aparajit-pratap @smangarole do we need versioning incrementing for a tool which we dont ship with Dynamo? My feeling is no.

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md (hunk 2 above)
Comment (mjkkirschner) [reply src/Tools/TransformXMLToJSONUsingXSLT/README.md]: 👍

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    <add key="textFile" value="CustomTextResult.txt" />
+    <add key="textFileWithFilters" value="TextResultWithFilters.txt" />
+    <add key="applyFilters" value="1" /><!--1 means yes, 0 means not-->
+</appSettings>
+```
+**About TestResult-Color_Tests.xml**
+This file is in format NUnit2 and it is the result of a set of running test, there is an example above:
+Enter to the path where the console of NUnit where installed by default C:\Program Files (x86)\NUnit.org\nunit-console and the type the next command.
+```
+nunit3-console.exe "C:\\DSCoreNodesTests.dll" --where "test=="DSCoreNodesTests.ColorTests" and cat != Failure and cat != BackwardIncompatible" --labels=Before --explore="C:\\TestResult-Color_Tests.xml";format=nunit2
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToJSONUsingXSLT/README.md]: I think you should *not* assume that nunit3 runner is installed - can you look at pulling it as part of the build:
something like this perhaps?
https://www.nuget.org/packages/NUnit.Console/

@alfredo-pozo @smangarole  @QilongTang please consider if this should be done in this PR or as part of the CI build job instead.

File: .gitignore (hunk 9 above)
Comment (manuelsaldivar525) [reply .gitignore]: Yes, I already reverted it.

File: src/Dynamo.All.sln (hunk 8 above)
Comment (manuelsaldivar525) [reply src/Dynamo.All.sln]: Yes, I already reverted it.

File: src/Tools/TransformXMLToJSONUsingXSLT/App.config (hunk 11 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToJSONUsingXSLT/App.config]: This reference to Newtonsoft.Json won´t be needed.

File: src/Tools/TransformXMLToJSONUsingXSLT/README.md (hunk 13 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToJSONUsingXSLT/README.md]: This was just a sample to show how can we get the test result file.

File: src/Tools/TransformXMLToJSONUsingXSLT/Properties/AssemblyInfo.cs (hunk 12 above)
Comment (aparajit-pratap) [reply src/Tools/TransformXMLToJSONUsingXSLT/Properties/AssemblyInfo.cs]: Not necessary.

File: src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs [hunk 14]
```diff
@@ -1,67 +0,0 @@
...
-//
-//      Major Version
-//      Minor Version 
-//      Build Number
-//      Revision
-//
-// You can specify all the values or you can default the Build and Revision Numbers 
-// by using the '*' as shown below:
-// [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyFileVersion("2.2.0.4207")]
```
Comment (mjkkirschner) [inline src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs]: you need to revert this change - this file should not be removed.

File: src/Dynamo.All.sln [hunk 15]
```diff
@@ -512,7 +512,6 @@ Global
 		{F7FD9395-35D5-474D-8AD1-9904817E70D3}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
 		{F7FD9395-35D5-474D-8AD1-9904817E70D3}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{F7FD9395-35D5-474D-8AD1-9904817E70D3}.Release|Any CPU.ActiveCfg = Release|Any CPU
-		{F7FD9395-35D5-474D-8AD1-9904817E70D3}.Release|Any CPU.Build.0 = Release|Any CPU
```
Comment (mjkkirschner) [inline src/Dynamo.All.sln]: please try to revert this change - I don't think there should be any modifications to this .sln right?

File: src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs (hunk 14 above)
Comment (manuelsaldivar525) [reply src/AssemblySharedInfoGenerator/AssemblySharedInfo.cs]: I already reverted it, and I added it like it was at the beginning of this PR.

File: src/Dynamo.All.sln (hunk 15 above)
Comment (manuelsaldivar525) [reply src/Dynamo.All.sln]: I already reverted it.

File: src/Tools/TransformXMLToCSVTool/Program.cs [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        static string textFileWithFiltersSlowTests = ConfigurationManager.AppSettings["textFileWithFiltersSlowTests"];
+        static string textFileWithFiltersFastTests = ConfigurationManager.AppSettings["textFileWithFiltersFastTests"];
+        static List<String> allLines = new List<string>();
+        static List<String> restLines = new List<string>();
+
+        static void Main(string[] args)
+        {
+            XSLTToAnyType();
+        }
+        public static void XSLTToAnyType()
```
Comment (mjkkirschner) [inline src/Tools/TransformXMLToCSVTool/Program.cs]: @manuelsaldivar525 looking sold- please add ///summary tags to public methods explaining what the method does.

**Comment by mjkkirschner** [comment]
Thanks for the work on this one, one last comment and looks good to merge - the builds are all passing now - yay.

File: src/Tools/TransformXMLToCSVTool/Program.cs (hunk 16 above)
Comment (manuelsaldivar525) [reply src/Tools/TransformXMLToCSVTool/Program.cs]: Yes, I added documentation.

**Comment by mjkkirschner** [comment]
@manuelsaldivar525 @smangarole what is the deal with this PR? Can we close it?

**Comment by manuelsaldivar525** [comment]
> @manuelsaldivar525 @smangarole what is the deal with this PR? Can we close it?

I think you can close it unless Sid says something different.

**Comment by mjkkirschner** [comment]
closing this, we can always re-open.
